pymeasure/instruments/rohdeschwarz/fsl.py

```diff
@@ -0,0 +1,256 @@
+        return self.values("SWE:TIME?")[0]
+
+    @sweep_time.setter
+    def sweep_time(self, value):
+        if type(value) is str and value.upper() == "AUTO":
+            self.write("SWE:TIME:AUTO ON")
+        else:
+            self.write(f"SWE:TIME {value}")
+
+    continuous_sweep = Instrument.control(
+        "INIT:CONT?",
+        "INIT:CONT %s",
+        "Continuous (True) or single sweep (False)",
+        validator=strict_discrete_set,
+        values=[True, False],
+        set_process=lambda x: "ON" if x else "OFF",
+        get_process=lambda x: bool(x),
+    )
+
+    def single_sweep(self):
+        """Perform a single sweep with synchronization."""
+        self.write("INIT; *WAI")
+
+    def continue_single_sweep(self):
+        """Continue with single sweep with synchronization."""
+        self.write("INIT:CONM; *WAI")
+
+    # Traces ------------------------------------------------------------------
+
+    def read_trace(self, n_trace=1):
+        """
+        Read trace data.
+
+        :param n_trace: The trace number (1-6). Default is 1.
+        :return: Nnumpy arrays of the trace data (frequency and amplitude).
+        """
+        y = np.array(self.values("TRAC{n_trace}? TRACE{n_trace}"))
+        x = np.linspace(self.freq_start, self.freq_stop, len(y))
+        return x, y
+
+    trace_mode = Instrument.control(
+        "DISP:TRAC:MODE?",
+        "DISP:TRAC:MODE %s",
+        "Trace mode ('WRIT', 'MAXH', 'MINH', 'AVER' or 'VIEW')",
+        validator=strict_discrete_set,
+        values=["WRIT", "MAXH", "MINH", "AVER", "VIEW"],
+    )
+
+    # Markers -----------------------------------------------------------------
+
+    def create_marker(self, num=1, is_delta_marker=False):
+        """
+        Create a marker.
+
+        :param num: The marker number (1-4)
+        :param is_delta_marker: True if the marker is a delta marker, default
+            is False.
+        :return: The marker object.
+        """
+        return self.Marker(self, num, is_delta_marker)
+
+    class Marker:
+        def __init__(self, instrument, num, is_delta_marker):
+            """
+            Marker and Delta Marker class.
+
+            :param instrument: The FSL instrument.
+            :param num: The marker number (1-4)
+            :param is_delta_marker: True if the marker is a delta marker,
+                defaults to False.
+            """
+            self.instrument = instrument
+            self.is_delta_marker = is_delta_marker
+            # Building the marker name for the commands.
+            if self.is_delta_marker:
+                # Smallest delta marker number is 2.
+                self.name = "DELT" + str(max(2, num))
+            else:
+                self.name = "MARK"
+                if num > 1:
+                    # Marker 1 doesn't get a number.
+                    self.name = self.name + str(num)
+
+            self.activate()
+
+        def write(self, command):
```

Comment:
you have overridden `write`, but not the `read` method. If that is intentional (e.g. because the instrument does not include this CALC... prefix when replying), just leave a comment here so that future readers are not confused. :-)

Reply:
In 86744f8 I added `read` and `ask` and also fixed the `value` method (I didn't actually delegate to `instrument` causing an infinite loop. Now you can actually read out marker positions etc.

One thing that I noticed is that the [link you provided earlier](https://github.com/pymeasure/pymeasure/blob/7580edd2cc5dfaafb9c0d6462d6f95cd87a3d468/pymeasure/instruments/tektronix/afg3152c.py#L127-L144) doesn't return anything for the `read` and `ask` methods. Maybe this didn't get caught because `values` was always used by the other methods?

Reply:
Wow, nice catch about those methods not returning! That looks like an error - can you also include a fix while you're at it?

Reply:
I opened [another pull request](https://github.com/pymeasure/pymeasure/pull/519) since I also took the opportunity to fix some formatting issues.